### PR TITLE
Refactor models import issue13881

### LINKED ISSUE #13981: Refactor `models.py` API
### Description

Currently, `astropy.modeling.models` has a lot of `*` imports. Instead it should be refactored to have an explicit `__all__` and explicit imports.

### PR DESCRIPTION
**Pull request by Maia-jp**
Refactor `astropy.modeling.models` to use explicit imports and define an explicit `__all__` list.

Changes:
- Replaced the wildcard import (`from . import math_functions as math`) with an explicit import statement.
- Added explicit imports for the modules `core`, `functional_models`, `mappings`, `physical_models`, `polynomial`, `powerlaws`, `projections`, `rotations`, `spline`, and `tabular`.
- Created an `__all__` list that includes all the imported modules and specific model classes.

Pull Request Description:
This pull request refactors `astropy.modeling.models` to improve code organization and maintainability. The changes introduce explicit imports and define an explicit `__all__` list to clearly specify the exported symbols. Additionally, a docstring explaining constraints has been attached to models that support them.

Fixes #13981

### CODE REVIEW

**Comment by WilliamJamieson** [comment]
@Maia-jp, thanks for the contribution. Will you rebase your changes on the latest version of the main and run pre-commit?

Please let me know if you have any issues with doing this.

File: astropy/modeling/models.py [hunk 1]
```diff
@@ -6,15 +6,152 @@
...
+    Cosine1D,
+    Tangent1D,
+    ArcSine1D,
+    ArcCosine1D,
+    ArcTangent1D,
+    Trapezoid1D,
+    TrapezoidDisk2D,
+    Voigt1D,
+    KingProjectedAnalytic1D,
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:45]: @nstarman, you've been thinking more about the `__all__` API than me. Do you know of a more "elegant" solution to importing everything from the `__all__` of a submodule without using the `*`?

File: astropy/modeling/models.py [hunk 2]
```diff
@@ -6,15 +6,152 @@
...
+    Polynomial1D,
+    Polynomial2D
+)
+from .powerlaws import (
+    PowerLaw1D,
+    BrokenPowerLaw1D
+)
+from .projections import (
+    Pix2Sky_TAN
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:73]: Many of the test failures appear to stem from this `__all__` not being propagated here: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/projections.py#L67-L80

File: astropy/modeling/models.py [hunk 3]
```diff
@@ -6,15 +6,152 @@
...
+from .mappings import (
+    Identity,
+    Mapping,
+    UnitsMapping
+)
+from .physical_models import (
+    BlackBody,
+    NFW,
+    Drude1D
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:55]: It appears `Plummer1D` is missing here: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/physical_models.py#L18.

File: astropy/modeling/models.py [hunk 4]
```diff
@@ -6,15 +6,152 @@
...
+from .polynomial import (
+    SIP,
+    Chebyshev1D,
+    Chebyshev2D,
+    Hermite1D,
+    Hermite2D,
+    Legendre2D,
+    Legendre1D,
+    Polynomial1D,
+    Polynomial2D
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:65]: There are missing items from: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/polynomial.py#L18-L31

File: astropy/modeling/models.py [hunk 5]
```diff
@@ -6,15 +6,152 @@
...
+)
+from .powerlaws import (
+    PowerLaw1D,
+    BrokenPowerLaw1D
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:70]: There are missing items from: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/powerlaws.py#L13-L20

File: astropy/modeling/models.py [hunk 6]
```diff
@@ -6,15 +6,152 @@
...
+from .projections import (
+    Pix2Sky_TAN
+)
+from .rotations import (
+    Rotation2D,
+    RotateNative2Celestial,
+    RotateCelestial2Native,
+    EulerAngleRotation
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:79]: Missing things from: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/rotations.py#L33-L40

File: astropy/modeling/models.py [hunk 7]
```diff
@@ -6,15 +6,152 @@
...
+from .spline import (
+    Spline1D
+)
```
Comment (WilliamJamieson) [inline astropy/modeling/models.py:82]: There are missing things from: https://github.com/astropy/astropy/blob/5b771e331663f65bc07b75dd700774be7db37dd1/astropy/modeling/spline.py#L18-L24

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
In order for this PR to be backwards compatible with the `from foo import *` imports currently being used, it must import everything currently being imported via that mechanism.

Python determines what to import when using `from foo import *` via importing everything within the `__all__` defined for the `foo` module. If `foo` defines no `__all__` by default it is all of the objects whose name is not prefixed by `_` (ie. `_private`). See https://docs.python.org/3/tutorial/modules.html#importing-from-a-package for details.

File: astropy/modeling/models.py (hunk 1 above)
Comment (saimn) [reply astropy/modeling/models.py:45]: > Do you know of a more "elegant" solution to importing everything from the __all__ of a submodule without using the *?

Well, either you import explicitly a long list of things (duplicating another long list in the imported module's `__all__`), with the risk that the two lists will not be consistent, or you use `import *`. 
This is one of the topics of the Public API discussion. 
(Scipy kept the `import *` after making the submodules private:
https://github.com/scipy/scipy/blob/main/scipy/ndimage/__init__.py#L152-L156)

**Pull request closed without merging**